x/delayedack/ibc_middleware.go

```diff
@@ -163,7 +164,66 @@ func (im IBCMiddleware) OnAcknowledgementPacket(
 	acknowledgement []byte,
 	relayer sdk.AccAddress,
 ) error {
-	return im.app.OnAcknowledgementPacket(ctx, packet, acknowledgement, relayer)
+	if !im.keeper.IsRollappsEnabled(ctx) {
+		return im.app.OnAcknowledgementPacket(ctx, packet, acknowledgement, relayer)
+	}
+	logger := ctx.Logger().With("module", "DelayedAckMiddleware")
+
+	// no-op if the packet is not a fungible token packet
+	var data transfertypes.FungibleTokenPacketData
+	if err := transfertypes.ModuleCdc.UnmarshalJSON(packet.GetData(), &data); err != nil {
+		return err
+	}
+
+	// Check if the packet is destined for a rollapp
+	chainID, err := im.keeper.ExtractChainIDFromChannel(ctx, packet.DestinationPort, packet.DestinationChannel)
+	if err != nil {
+		logger.Error("Failed to extract chain id from channel", "err", err)
+		return err
+	}
+
+	_, found := im.keeper.GetRollapp(ctx, chainID)
+	if !found {
+		logger.Debug("Skipping IBC transfer OnAcknowledgementPacket for non-rollapp chain")
+		return im.app.OnAcknowledgementPacket(ctx, packet, acknowledgement, relayer)
+	}
+
+	// Get the light client height at this block height as a proxy for the packet proof height
+	clientState, err := im.keeper.GetClientState(ctx, packet)
+	if err != nil {
+		return err
+	}
+
+	// TODO(omritoptix): Currently we use this height as the proofHeight as the real proofHeight from the ibc lower stack is not available.
+	// using this height is secured but may cause extra delay as at best it will be equal to the proof height (but could be higher).
+	ibcClientLatestHeight := clientState.GetLatestHeight()
+	finalizedHeight, err := im.keeper.GetRollappFinalizedHeight(ctx, chainID)
+	if err == nil && finalizedHeight >= ibcClientLatestHeight.GetRevisionHeight() {
+		logger.Debug("Skipping IBC transfer OnAcknowledgementPacket as the packet proof height is already finalized")
+		return im.app.OnAcknowledgementPacket(ctx, packet, acknowledgement, relayer)
+	}
+	// Run the underlying app's OnAcknowledgementPacket callback
+	// with cache context to avoid state changes and report the acknowledgement result.
+	// Only save the packet if the underlying app's callback succeeds.
+	cacheCtx, _ := ctx.CacheContext()
```

Comment:
why u need it? u already run it with cacheTx by the caller

Reply:
Yea but than the caller controls it.
after it succeeds with cache context, it will actually run it without - which will make the ack run before the packet is finalized.
I want to force it to always run with cache context by the caller and run it with non-cached context only from the hook. 
Maybe there is a better way to do it?

Reply:
oh I see what u did here
what about the case err != nil?
we don't want to write the cacheCtx?

Reply:
can you elaborate? not sure what you mean by writing the cacheCtx.